George temp

thezeit opened the pull request:
1. This version creates pickle files containing dicts of vocab (2 versions: {word{word_types[word_type_defs]}} and the other {word_types{all vocab of that type}}
2. Game engine now uses these pickle files to run
3. user_input handles user input more precisely
4. Utilises set_choose module from NG
5. Other general cleaning up inlcuding reducing api calls

thezeit committed:
Made some basic changes based on Eoin's comments,
including more specific error handling and
reducing api calls.

thezeit committed:
user_vocab altered for better Exception handling
and to incoporate NormalGaussian edits

set_choose added as created by NormalGaussian

vocab_game - possible line movement (no
substantive changes)

thezeit committed:
vocab_game has improved exception handling.
Removed unneccessaery print statements from
get_defs and user_vocab.
Not fixed: loop and main game engine. Plan to use
new user vocabs data structures for
accessing vocab so will do this
first.
For future: Possible future update to include
seperate user input handling module.

thezeit committed:
Created seperate pickle dictionary for storing
vocab. This is intended to be the basis for a new
game engine in future commits.

thezeit committed:
Primary changes are to the game engine which
now uses the pickle dict and the
addition of more advanced user interaction in
user_input.

Game now works at acceptable speed.

NormalGaussian commented:
Attempted to run using `python3 ./game_engine.py`; got:

```bash
eoin@Marasi:~/Coding/vocab-builder$ python3 ./game_engine.py 
Traceback (most recent call last):
  File "/home/eoin/Coding/vocab-builder/./game_engine.py", line 16, in <module>
    usr_vocab_def = (unpickle_vocab('known_vocab.pickle')[0])
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/eoin/Coding/vocab-builder/pickle_vocab.py", line 90, in unpickle_vocab
    with open(location, 'rb') as file:
         ^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: 'known_vocab.pickle'
```

NormalGaussian commented:
Can't quickly see how to generate the required pickles; pausing review until either (or both):
- bug resolved
- README updated with instructions on how to run

/label invalid

@thezeit

NormalGaussian commented on set_choose.py:
```diff
@@ -0,0 +1,28 @@
+T = TypeVar('T')
+
+
+#To note - the index reference is not constant
+#due to set
```
This comment would be better as a documentation string

NormalGaussian commented on set_choose.py:
```diff
@@ -0,0 +1,28 @@
+#To note - the index reference is not constant
+#due to set
+def setChoice(s: Set[T]) -> T | None:
+    if len(s) == 0:
```
```suggestion
def setChoice(s: Set[T]) -> T | None:
    """
    Chooses a single element at random from a set.
    When given the empty set returns None.
    When given a non-empty set always returns a value.
    """
    if len(s) == 0:
```

An example doc string. If this is applied, it will make it so hovering over the function name at point of use (in another file) will show the doc in the popup.

If you want to make a note about the implementation choice of a function it is common to do it as a comment after the docstring.

NormalGaussian commented on set_choose.py:
```diff
@@ -0,0 +1,28 @@
+#due to set
+def setChoice(s: Set[T]) -> T | None:
+    if len(s) == 0:
+        return None
```
nb. It may be desirable to raise an exception here rather than return none.

NormalGaussian commented on user_vocab.py:
```diff
@@ -1,21 +1,47 @@
+if filename.endswith('.txt'):
+    pass
+else:  
+    raise Exception("user_vocab parsing curently unsupported file type")
```
A note about files; windows is the only OS that particularly cares about file extensions - and even then they don't strictly matter. it is the practice of *NIX first programs actively view the filepath/filename as extraneous metadata - not least because files can be streams (complex topic).

The point being, in many cases checking file extension can be overly restrictive. As such I would typically encourage warnings over errors unless there was some major risk involved.

That said, in this case we are fairly certain we are parsing a particular file from a kindle - so the added restriction is very valid. Under that context the error message could do with some work; maybe communicating that we only currently parse kindle highlights files - "unable to proceed, expected filename to end in .txt. Expected file to be a kindle highlights file, which ends in .txt" is verbose and needs work but not vague.

NormalGaussian commented on user_vocab.py:
```diff
@@ -1,21 +1,47 @@
-else:
-    with open(filename) as f:
-        content = f.read().splitlines()
+    raise Exception('File does not exist.')
```
Its important when raising errors not to lie or mislead; in this case we the error is, in the strictest sense, accurate but may mislead.

If the file is actually a directory or block device we would also get this error.

Clarity in error handling is key, its also a good idea to communicate *which* file we can't find.

e.g. `f"Cannot process {filename}; it either does not exist or is not a file"`

NormalGaussian commented on user_vocab.py:
```diff
@@ -1,21 +1,47 @@
+    content = f.read().splitlines()
 
+#removes any characters that are not letters ('abc')
 def rmv_non_charac(word):
```
```suggestion
def remove_non_alpha_characters(word):
```

I found your missing vowels :P

Also - all possible components of a string are "characters" inlcuding "non-printing characters" and "control characters". https://en.wikipedia.org/wiki/Control_character

The set `[a-zA-Z]` is the [latin] alphabet commonly shortened to alpha. When including numbers `[a-zA-Z0-9]` we say alphanumeric.

NormalGaussian commented on user_input.py:
```diff
@@ -0,0 +1,59 @@
+from time import sleep
+from random import choice
+
+savage1 = {"ducky": "Has feathers and goes quack quack", "red": "The colour your face should be", "cup": "sippy sippy", "feeble": "An assessment of your mental prowess..."}
```
?

NormalGaussian commented on user_input.py:
```diff
@@ -0,0 +1,59 @@
+from textwrap import dedent
+from time import sleep
+from random import choice
```
```suggestion
from time import sleep
```

Removes unused imports

NormalGaussian commented on user_input.py:
```diff
@@ -0,0 +1,59 @@
+from random import choice
+
+savage1 = {"ducky": "Has feathers and goes quack quack", "red": "The colour your face should be", "cup": "sippy sippy", "feeble": "An assessment of your mental prowess..."}
+
```
As a whole the content of this file seems tightly coupled with game_engine in a confusing way- it has assumptions from game_engine (the number of options, the state of user-interaction) and its not got a clear domain. I'm assuming you haven't decided how you want to arrange this, so keep an eye on it.

NormalGaussian commented on user_input.py:
```diff
@@ -0,0 +1,59 @@
+savage1 = {"ducky": "Has feathers and goes quack quack", "red": "The colour your face should be", "cup": "sippy sippy", "feeble": "An assessment of your mental prowess..."}
+
+
+def check_input(user_input):
```
This function has a return signature of `None | True` - you probably want `Boolean` aka `True | False`.

NormalGaussian replied:
```suggestion
def check_input(user_input: string) -> bool:
```

I think

NormalGaussian commented on user_input.py:
```diff
@@ -0,0 +1,59 @@
+
+
+
+def check_response(wrong, conc_wrong):
```
The name of the function and its behaviour live in different conceptual domains!

The name `check_response` implies some form of validation and returning a boolean / object extending boolean concepts. However the behaviour is to return `None` - ie. not to be used for validation - and to interact with the user.

This makes reading and debugging and extending code harder and more confusing.

NormalGaussian commented on get_defs.py:
```diff
@@ -5,40 +5,45 @@ def api_call(word: str) -> list:
     
     trial = 0
     time.sleep(0.5)  #0.4 sleep with 2 attempts has had best results
```
we'll come back to this later; how to best handle a rate limited api / how its commonly handled

NormalGaussian commented on get_defs.py:
```diff
@@ -5,40 +5,45 @@ def api_call(word: str) -> list:
+                raise Exception(f"non-200 status code returned for {word}: {response.status_code}")
+            
+
+        except Exception:
             print(f"Bad Request - no relvant url found for {word}")
             return None
```
2 things;

1. this message lies!
2. the return of None is probably problematic as to higher level control flow.

thezeit replied:
I'm currently trying (bool) False instead. Possible issue being 0 == False so let me know if there's a standard/norm to follow

NormalGaussian commented on get_defs.py:
```diff
@@ -5,40 +5,45 @@ def api_call(word: str) -> list:
-                return None
-    except:
+
+    except IndexError:
+        print(f"Index Error raised in word_types for {word}.")
         return None
```
1. I really want to get you away from control-flow using None for exceptional behaviour as quickly as I can.

2. Note the error can be captured and printed using `except IndexError as indexError:` and then printed, or its message directly accessed for formatting in another error. The error can even be re-raised.

3. An IndexError is the kind of error which *practically never* has to occur - if a length check is done there should never be an IndexError raised. The only exception would be somehow not having access to the size or removing the check for extreme performance reasons.

NormalGaussian commented on get_defs.py:
```diff
@@ -5,40 +5,45 @@ def api_call(word: str) -> list:
         return None
+    
+    else:
+        print(f"Non-index error raised in word_types for {word}.")
```
There is a consistency discrepancy here.

As this is a returning function, it is the done thing to never not return. Failing to return implies that additional logic can safely be added after the statement, which isn't the case here as this is raising an error.

Additionally, `else` is called when no exception is raised in the try block (its not commonly used). I think you meant `except:` to handle every raised value that isn't an IndexError.

thezeit replied:
This has now been altered to:

`def word_types(word):
    
    """Returns a set of all found vocab types for a given word from 
    api (e.g. {"verb", "noun"})
    """

    word_types = set()
    current_call = api_call(word)

    if current_call != False:
        for items in current_call:
            for mean_items in items['meanings']:
                word_types.add(mean_items['partOfSpeech'])
        if len((word_types)) > 0:
            return word_types

    else:
        print("Should never run.")`

thezeit committed:
Allowed creation of pickle files automatically
from the running of game_engine
Also resolved issue cause by a pickle function
returning a tuple

NormalGaussian commented on get_defs.py:
```diff
@@ -62,13 +67,10 @@ def extract_defs(word, vocab_type = None):
     #Returns all definiitions of all types
-    elif api_call(word) != None:
-        for items in api_call(word):
+    elif current_call != None:
+        for items in current_call:
             for mean_items in items['meanings']:
                 for spec_defs in mean_items['definitions']:
                     all_defs.append(spec_defs['definition'])
         if len(all_defs) > 0:
             return all_defs
-    
-    return None
```
1. consistency in return

2. naming can make this a lot more readable. Imagine something like "lookup_word" over api_call, and "result" over "current_call". Be careful of phrasing like "current_call" which imply the api call is ongoing.

The abstraction to a single api_call has reaped significant performance benefits!

thezeit committed:
YUpdated git ignore

NormalGaussian reviewed (changes requested):
Good progression; many notes on areas in which you are stretching yourself. I hope its helpful.

NormalGaussian commented:
Bug?

```
Your score is currently 0. You require 5 points to leave and continue with your life...

"(interrogative) At what time? At which time? Upon which occasion or circumstance? Used to introduce direct or indirect questions about time."

To which of the following options does this definition pertain?
      chary
      incongruously
      agog
      wen

Please enter an integer from 1 - 4
>>> wen
User, you were given precise instructions.
Integers 1 - 4 only.
Let's try again shall we...
Your score is currently 0. You require 5 points to leave and continue with your life...

"At an earlier time and under different, usually less favorable, circumstances."

To which of the following options does this definition pertain?
      chary
      wen
      incongruously
      assiduously

Please enter an integer from 1 - 4
```

It seems that the agog was dropped in favour of assiduously when representing the question

NormalGaussian commented:
On the first question I entered

```
7
7
7
7
7
7
```

ie. pressing enter between each one very quickly.

This has made it gradually iterate through the list of numbers, stopping me from playing the game.

NormalGaussian commented:
note: it would be helpful to see the number against each word

NormalGaussian commented:
Colour would make it easier to see success and failure, as well as locate the relevant part of the question / answers.

NormalGaussian commented:
The printing is a little slow once you get into the game

NormalGaussian commented:
It would be super awesome if the screen cleared between questions

NormalGaussian commented:
when downloading the words, it would be useful to see how many words its downloading and its current progress rather than just the errors

thezeit committed:
Add docstring to set choice

Co-authored-by: Normal Gaussian <48660366+NormalGaussian@users.noreply.github.com>

thezeit committed:
Update user_vocab.py

Co-authored-by: Normal Gaussian <48660366+NormalGaussian@users.noreply.github.com>

thezeit committed:
Incorporated minor changes from accepting NG
suggestions on github (in broswer) to local

thezeit commented:
> It would be super awesome if the screen cleared between questions

Resolved

thezeit committed:
Several quick fixes and alterations made based
on NG comments including altering return from
None to flase in several places, indices
print adjacent to answer and renaming.

thezeit committed:
Deleted printing lines that tracked "url not
found" and others whilst running lookup for
pickle. Now prints progress every 10 iterations
(i.e. words processed)

thezeit committed:
Rebuilt the game engine to give more specificity
to its functions, and those that it calls. Solved
the input stream bug, the repetition issue, the
altered question bug after "User, you were
given..." behaviour. Also introduced typing speed
for printing to the user under certain conditions.
Also general doc-string alterations for clarity.

thezeit committed:
Added a basic check to see whether enough words
exist for a particular option (are there enough
nouns for example). Also added colour to console
output.

thezeit committed:
Minor edit made and func deleted. Quick save
before more substantive changes are made.

thezeit committed:
Addition of colour for the answer when wrong.
Observed unresolved bug where 3 answers appeared
instead of 4. Tracked on github.

thezeit committed:
Made alterations to the pickle file (and functions
that read it) to store counts for concurrent right
and wrong answers for words (for each version by
type) within the known_vocab.pickle file.

thezeit committed:
Now includes functions which add/subtract to
concurrent corect/wrong. However this state is
not yet stored.

egroat committed:
Corrects code for clearing stdin on linux/unix

egroat committed:
Renamed clear_standard_in to clear_stdin to follow convention

egroat committed:
Proposes changing 'concurrent_wrong' to 'consecutive_incorrect_answers'

concurrent is an incorrect and misleading term for what is being tracked

egroat committed:
Alters naming of function 'generate_question' to reflect its behaviour 'generate_and_ask_question'

Also updates docstring

egroat committed:
Changes variable 'wrong: False | int' to 'total_incorrect_answers: int'

egroat committed:
Carries forward variable renaming into child functions

egroat committed:
Removes unused imports and constants

egroat committed:
Removes unnecessary extra variable was_wrong

thezeit committed:
Merge pull request #10 from NormalGaussian/fix/9-clear-standard-in-linux

Corrects code for clearing stdin on linux/unix

egroat committed:
Updates function 'check_response' to have a name that matches its behaviour

check_response would be assumed to return True / False, some form of validation object, or raise an exception.

It would not be expected to output berations to the user

thezeit committed:
Merge pull request #12 from NormalGaussian/fix/minor-changes

Fix/minor changes

thezeit committed:
Merge branch 'george-temp' of https://github.com/NormalGaussian/vocab-builder into george-temp

thezeit committed:
Now includes the abiltiy to save player progress
in "working_vocab.pickle" for future games. player
can now remove words from their working dict.
"known_vocab.pickle" now functions as a master
copy.

thezeit committed:
Minor docstring update

thezeit committed:
Fixed bugs #14 and #13 found by NG.
Another issue found due to api calls using VPN.
Solved by switching off.
